src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs

```diff
@@ -14,211 +24,3020 @@ namespace Microsoft.CodeAnalysis.CSharp
         {
-            VisitStatement(finallyBlock); // this should generate no pending branches
+            Debug.Assert(!_trackExceptions || this.PendingBranches.Count > 0 && this.PendingBranches[0].Branch == null);
+            var result = Visit(node);
+            Unsplit();
+            return result;
         }
 
-        #endregion TryStatements
+        /// <summary>
+        /// Visit a statement.
+        /// </summary>
+        [DebuggerHidden]
+        protected virtual void VisitStatement(BoundStatement statement)
+        {
+            Debug.Assert(!_trackExceptions || this.PendingBranches.Count > 0 && this.PendingBranches[0].Branch == null);
+            Visit(statement);
+            Debug.Assert(!this.IsConditionalState);
+        }
+
+        protected static bool IsConstantTrue(BoundExpression node)
+        {
+            return node.ConstantValue == ConstantValue.True;
+        }
+
+        protected static bool IsConstantFalse(BoundExpression node)
+        {
+            return node.ConstantValue == ConstantValue.False;
+        }
+
+        protected static bool IsConstantNull(BoundExpression node)
+        {
+            return node.ConstantValue == ConstantValue.Null;
+        }
+
+        /// <summary>
+        /// Called at the point in a loop where the backwards branch would go to.
+        /// </summary>
+        private void LoopHead(BoundLoopStatement node)
+        {
+            TLocalState previousState;
+            if (_loopHeadState.TryGetValue(node, out previousState))
+            {
+                Join(ref this.State, ref previousState);
+            }
+
+            _loopHeadState[node] = this.State.Clone();
+        }
+
+        /// <summary>
+        /// Called at the point in a loop where the backward branch is placed.
+        /// </summary>
+        private void LoopTail(BoundLoopStatement node)
+        {
+            var oldState = _loopHeadState[node];
+            if (Join(ref oldState, ref this.State))
+            {
+                _loopHeadState[node] = oldState;
+                this.stateChangedAfterUse = true;
+            }
+        }
+
+        /// <summary>
+        /// Used to resolve break statements in each statement form that has a break statement
+        /// (loops, switch).
+        /// </summary>
+        private void ResolveBreaks(TLocalState breakState, LabelSymbol label)
+        {
+            var pendingBranches = PendingBranches;
+            var count = pendingBranches.Count;
+
+            if (count != 0)
+            {
+                int stillPending = 0;
+                for (int i = 0; i < count; i++)
+                {
+                    var pending = pendingBranches[i];
+                    if (pending.Label == label)
+                    {
+                        Join(ref breakState, ref pending.State);
+                    }
+                    else
+                    {
+                        if (stillPending != i)
```

Comment:
when 'i' is being used for more than simple indexing, it's useful to give it a clarifying name to help indicate what it means.